mooncake-transfer-engine/src/topology.cpp

```diff
@@ -0,0 +1,205 @@
+#include <glog/logging.h>
+#include <jsoncpp/json/json.h>
+
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+#ifdef USE_CUDA
+#include "cuda_runtime.h"
+#endif
+
+#include <ctype.h>
+#include <dirent.h>
+#include <limits.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+
+#include "topology.h"
+
+struct InfinibandDevice {
+    std::string name;
+    std::string pci_bus_id;
+    int numa_node;
+};
+
+struct TopologyEntry {
+    std::string name;
+    std::vector<std::string> preferred_hca;
+    std::vector<std::string> avail_hca;
+
+    Json::Value to_json() {
+        Json::Value matrix(Json::arrayValue);
+        Json::Value hca_list(Json::arrayValue);
+        for (auto &hca : preferred_hca) {
+            hca_list.append(hca);
+        }
+        matrix.append(hca_list);
+        hca_list.clear();
+        for (auto &hca : avail_hca) {
+            hca_list.append(hca);
+        }
+        matrix.append(hca_list);
+        return matrix;
+    }
+};
+
+static std::vector<InfinibandDevice> list_infiniband_devices() {
+    DIR *dir = opendir("/sys/class/infiniband");
+    struct dirent *entry;
+    std::vector<InfinibandDevice> devices;
+
+    if (dir == NULL) {
+        LOG(WARNING) << "failed to list /sys/class/infiniband";
+        return {};
+    }
+    while ((entry = readdir(dir))) {
+        if (entry->d_name[0] == '.') {
+            continue;
+        }
+
+        std::string device_name = entry->d_name;
+
+        char path[PATH_MAX];
+        char resolved_path[PATH_MAX];
+        snprintf(path, sizeof(path), "/sys/class/infiniband/%s/../..",
```

Comment:
Is it equivant to `snprintf(path, sizeof(path), "/sys/class/");`?

Reply:
Nope, `/sys/class/infiniband/mlx5_X` is a symlink pointing to `/sys/devices/pciXXXX:XX/XXXX:XX:XX.X/infiniband/mlx5_X`, hence the realpath of `/sys/class/infiniband/%s/../..` is the PCI bus id.

Reply:
Maybe add a comment in code is necessary to avoid incorrect change in the future.